zcash_primitives/src/transaction/mod.rs

```diff
@@ -228,39 +302,18 @@ impl Transaction {
     pub fn read<R: Read>(reader: R) -> io::Result<Self> {
         let mut reader = HashReader::new(reader);
 
-        let header = reader.read_u32::<LittleEndian>()?;
-        let overwintered = (header >> 31) == 1;
-        let version = header & 0x7FFFFFFF;
-
-        let version_group_id = if overwintered {
-            reader.read_u32::<LittleEndian>()?
-        } else {
-            0
-        };
-
-        let is_overwinter_v3 = overwintered
-            && version_group_id == OVERWINTER_VERSION_GROUP_ID
-            && version == OVERWINTER_TX_VERSION;
-        let is_sapling_v4 = overwintered
-            && version_group_id == SAPLING_VERSION_GROUP_ID
-            && version == SAPLING_TX_VERSION;
+        let version = TxVersion::read(&mut reader)?;
+        let is_overwinter_v3 = version == TxVersion::Overwinter;
+        let is_sapling_v4 = version == TxVersion::Sapling;
 
         #[cfg(feature = "zfuture")]
-        let has_tze = overwintered
-            && version_group_id == ZFUTURE_VERSION_GROUP_ID
-            && version == ZFUTURE_TX_VERSION;
+        let has_tze = version == TxVersion::ZFuture;
         #[cfg(not(feature = "zfuture"))]
         let has_tze = false;
 
-        if overwintered && !(is_overwinter_v3 || is_sapling_v4 || has_tze) {
-            return Err(io::Error::new(
-                io::ErrorKind::InvalidInput,
-                "Unknown transaction format",
-            ));
-        }
```

Comment:
@daira This prior code rejects all not-explicitly-valid combinations for Overwintered transactions, as `zcashd` does.